In [ ]:
!pip install pandas -q


#Importation des bibliothèques nécessaires
import pandas as pd
from pathlib import Path        


In [ ]:
#Ouvertude des premires datasets

In [ ]:
#Ouvertude / chargements des premires datasets


CountrySeries1= Path("/Users/thibaultdautheville/Downloads/Projet+Python_Dataset_Edstats_csv/EdStatsCountry-Series.csv")
StatsCountry2= Path("/Users/thibaultdautheville/Downloads/Projet+Python_Dataset_Edstats_csv/EdStatsCountry.csv")
StatsData3= Path("/Users/thibaultdautheville/Downloads/Projet+Python_Dataset_Edstats_csv/EdStatsData.csv")
StatsFootNote4= Path("/Users/thibaultdautheville/Downloads/Projet+Python_Dataset_Edstats_csv/EdStatsFootNote.csv")
StatsSeries5= Path("/Users/thibaultdautheville/Downloads/Projet+Python_Dataset_Edstats_csv/EdStatsSeries.csv")

CountrySeries1= pd.read_csv(CountrySeries1)
StatsCountry2= pd.read_csv(StatsCountry2)
StatsData3= pd.read_csv(StatsData3)
StatsFootNote4= pd.read_csv(StatsFootNote4)
StatsSeries5= pd.read_csv(StatsSeries5)


## Première partie : Analyse initiale des datasets – World Bank EdStats


---

### Détails des datasets

### 1. CountrySeries1  
**Shape : (613, 4)**  

Colonnes :  
- CountryCode  
- SeriesCode  
- DESCRIPTION  
- Unnamed  

Il s'agit d'une table de correspondance entre pays et indicateur, avec une description associée.

Cette table permet d’identifier, pour un pays donné et un indicateur donné, la source ou la spécificité méthodologique utilisée.

### Utilité :
- Vérifier la qualité des sources par pays  
- Détecter des différences méthodologiques  
- Justifier d’éventuels écarts entre pays  

---

### 2. StatsCountry2  
**Shape : (241, 32)**  

Il s'agit de la Dimension Pays.

Chaque ligne correspond à un pays, une région ou un groupe de pays.

Ce dataset contient des informations générales sur les pays :

- Données économiques  
- Groupe de revenu (Income Group)  
- Région  
- Monnaie  
- Standards IMF  
- Données sur les derniers recensements  
- Informations agricoles, industrielles, commerciales  
- Données administratives et statistiques  

À noter :  
Il y a 241 régions/pays/groupes de pays selon la classification habituelle, incluant des agrégats comme Arab World, Sub-Saharan Africa, etc.

#### Utilité :
- Analyse fine par groupe de pays  
- Segmentation par niveau de revenu  
- Construction de clusters  
- Croisement avec les données de population et d’éducation  

C’est un dataset clé pour l’analyse stratégique.

---

### 3. StatsData3  
**Shape : (886930, 70)**  

Le troisième dataset est le plus volumineux.

Il contient les données principales relatives :

- À la population (par tranche d’âge)  
- Au niveau d’éducation (par diplôme, genre, etc.)  
- Aux dépenses  
- À d’autres indicateurs éducatifs  

#### Structure :

Chaque ligne correspond à :  
1 pays × 1 indicateur  

On observe :  
241 pays × 3665 indicateurs ≈ 886 930 observations  

Les colonnes années vont de 1970 à 2100.

- 1970–2017 : données historiques  
- 2020–2100 : projections  
- Les années 2018–2019 sont absentes (probablement lié à la structure du dataset)

Cela représente une série temporelle de :

- 46 années historiques  
- 18 années projetées  
- Soit 64 points temporels  

Chaque colonne correspond à une année et contient la valeur de l’indicateur.

#### Grandes familles d’indicateurs :

1. Education  
2. Population  
3. Dépenses  

Ce dataset constitue le cœur de la matrice des données pour notre étude.

Il est relié aux autres dimensions :

- Pays → StatsCountry2  
- Indicateurs → StatsSeries5  
- Notes → StatsFootNote4  

Il présente un niveau de granularité extrêmement fin.

---

### 4. StatsFootNote4  
**Shape : (643637, 5)**  

Ce dataset contient les métadonnées spécifiques par observation.

Il s’agit des notes de bas de page associées à :  
1 pays + 1 indicateur + 1 année  

On observe environ 239 codes pays distincts.

Ce fichier apporte un complément sur :

- Les sources  
- Les estimations  
- Les précisions méthodologiques  
- Les particularités nationales  

Il est important pour qualifier la fiabilité des données.

---

### 5. StatsSeries5  
**Shape : (3665, 21)**  

Il correspond à la Dimension Indicateur.

Chaque ligne représente un indicateur unique.

Ce dataset permet de comprendre :

- Ce que mesure réellement l’indicateur  
- Sa définition courte et longue  
- Son thème (Topic)  
- Sa source  
- Sa méthodologie  
- Son unité de mesure  

Il est essentiel pour interpréter correctement les valeurs présentes dans StatsData3.

---

## Synthèse générale


Nous disposons :

- De 241 pays / régions  
- De 3665 indicateurs uniques  
- De données couvrant la période 1970–2100  
- D’un total de 886 930 combinaisons pays × indicateurs  

La structure globale correspond à un modèle en étoile (star schema) :

StatsSeries5 (Dimension Indicateur)  
        ↑  
StatsCountry2 ← StatsData3 → StatsFootNote4  
        ↓  
CountrySeries1  

StatsData3 est la table centrale autour de laquelle s’articulent les dimensions pays, indicateurs et notes.


In [ ]:

#Récupération des noms de colonnes et des dimensions de chaque DataFrame

names = [
    "CountrySeries1","StatsCountry2", "StatsData3", "StatsFootNote4", "StatsSeries5",
]

for name in names:
    df = globals().get(name)
    if isinstance(df, pd.DataFrame):
        print(f"{name} — shape: {df.shape}")
        print(df.columns.tolist())
        print()
    else:
        print(f"{name} introuvable ou n'est pas un DataFrame.\n")

CountrySeries1 — shape: (613, 4)
['CountryCode', 'SeriesCode', 'DESCRIPTION', 'Unnamed: 3']

StatsCountry2 — shape: (241, 32)
['Country Code', 'Short Name', 'Table Name', 'Long Name', '2-alpha code', 'Currency Unit', 'Special Notes', 'Region', 'Income Group', 'WB-2 code', 'National accounts base year', 'National accounts reference year', 'SNA price valuation', 'Lending category', 'Other groups', 'System of National Accounts', 'Alternative conversion factor', 'PPP survey year', 'Balance of Payments Manual in use', 'External debt Reporting status', 'System of trade', 'Government Accounting concept', 'IMF data dissemination standard', 'Latest population census', 'Latest household survey', 'Source of most recent Income and expenditure data', 'Vital registration complete', 'Latest agricultural census', 'Latest industrial data', 'Latest trade data', 'Latest water withdrawal data', 'Unnamed: 31']

StatsData3 — shape: (886930, 70)
['Country Name', 'Country Code', 'Indicator Name', 'Indicator C

In [ ]:
# Faire apparaître les deux premières lignes de chaque DataFrame

datasets = {
    'CountrySeries1': CountrySeries1,
    'StatsCountry2': StatsCountry2,
    'StatsData3': StatsData3,
    'StatsFootNote4': StatsFootNote4,
    'StatsSeries5': StatsSeries5
}

for dataset_name, df in datasets.items():
    print(f"\n{'='*80}")
    print(f"DATASET: {dataset_name}")
    print(f"{'='*80}\n")
    
    for row_idx in range(min(2, len(df))):
        print(f"\n--- Row {row_idx} ---")
        for col_name in df.columns:
            cell_value = df.loc[row_idx, col_name]
            print(f"{col_name}: {cell_value}")


DATASET: CountrySeries1


--- Row 0 ---
CountryCode: ABW
SeriesCode: SP.POP.TOTL
DESCRIPTION: Data sources : United Nations World Population Prospects
Unnamed: 3: nan

--- Row 1 ---
CountryCode: ABW
SeriesCode: SP.POP.GROW
DESCRIPTION: Data sources: United Nations World Population Prospects
Unnamed: 3: nan

DATASET: StatsCountry2


--- Row 0 ---
Country Code: ABW
Short Name: Aruba
Table Name: Aruba
Long Name: Aruba
2-alpha code: AW
Currency Unit: Aruban florin
Special Notes: SNA data for 2000-2011 are updated from official government statistics; 1994-1999 from UN databases. Base year has changed from 1995 to 2000.
Region: Latin America & Caribbean
Income Group: High income: nonOECD
WB-2 code: AW
National accounts base year: 2000
National accounts reference year: nan
SNA price valuation: Value added at basic prices (VAB)
Lending category: nan
Other groups: nan
System of National Accounts: Country uses the 1993 System of National Accounts methodology.
Alternative conversion factor: nan


##Schéma de l'architecture globale des données (sortie d'un LLM)


![Schéma](image%20schéma/mon_image.png)
![Schéma](image%20schéma/mon_image.png)


### Deuxième partie : début du nettoyage

Partie du code qui correspond à l'étape 2 du cours :
- calcul du nombre de doublons dans le jeu de données, suppression si il y en a
- Calcule de la proportion de valeurs manquantes par colonne et par jeux de données afin de supprimer les valeurs inutilisable
- Determiner la forme de chaque séries : valeur numérique (continue), ou valeur catégorielle
- Determiner les définition de chaque colonnes de chaque datasets

- Traitement de chaque fichier l'un après l'autre

#### Dataset 1 : CountrySeries

In [ ]:
# nombre de valeurs manquantes par colonne dans CountrySeries1 en pourcentage
CountrySeries1.isna().sum() / len(CountrySeries1) * 100

CountryCode      0.0
SeriesCode       0.0
DESCRIPTION      0.0
Unnamed: 3     100.0
dtype: float64

In [ ]:
#Connaître le type de données de chaque colonne dans CountrySeries1
CountrySeries1.dtypes

CountryCode        str
SeriesCode         str
DESCRIPTION        str
Unnamed: 3     float64
dtype: object

In [ ]:
CountrySeries1.duplicated().sum() #Connaitre le nombre de doublons

np.int64(0)

##### Résumé CountrySeries1 (613, 4)

attention codeMarkdown généré par LLM; non automatique
| Colonne      | Description | % de NA | Type de données |
|-------------|------------|---------|----------------|
| CountryCode | Code ISO 3 lettres du pays (clé de jointure). | 0.0 % | Texte (string / object) |
| SeriesCode  | Code unique de l’indicateur. | 0.0 % | Texte (string / object) |
| DESCRIPTION | Description spécifique liée au pays et à l’indicateur (source ou particularité méthodologique). | 0.0 % | Texte (string / object) |


In [ ]:

CountrySeries1 = CountrySeries1.drop(columns=['Unnamed: 3']) #supprimer la colonne "Unnamed: 3" qui ne contient que des valeurs manquantes


#Calculs automatiquement vérifié avec le viewer de DataFrame de VSCode 

KeyError: "['Unnamed: 3'] not found in axis"

#### Dataset 2: StatsCountry2


Mêmes opérations avec le deuxième datasets


In [ ]:
StatsCountry2.isna().sum() / len(StatsCountry2) * 100

Country Code                                           0.000000
Short Name                                             0.000000
Table Name                                             0.000000
Long Name                                              0.000000
2-alpha code                                           1.244813
Currency Unit                                         10.788382
Special Notes                                         39.834025
Region                                                11.203320
Income Group                                          11.203320
National accounts base year                           14.937759
National accounts reference year                      86.721992
SNA price valuation                                   18.257261
Lending category                                      40.248963
Other groups                                          75.933610
System of National Accounts                           10.788382
Alternative conversion factor           

In [ ]:
StatsCountry2.dtypes

Country Code                                             str
Short Name                                               str
Table Name                                               str
Long Name                                                str
2-alpha code                                             str
Currency Unit                                            str
Special Notes                                            str
Region                                                   str
Income Group                                             str
WB-2 code                                                str
National accounts base year                              str
National accounts reference year                     float64
SNA price valuation                                      str
Lending category                                         str
Other groups                                             str
System of National Accounts                              str
Alternative conversion f

In [ ]:
##Data set 2


StatsCountry2 = StatsCountry2.drop(columns=['WB-2 code'])


In [ ]:

#suppression de la colonne WB-2 _code qui ne contient qu'une copie des valeurs pays
StatsCountry2 = StatsCountry2.drop(columns=['WB-2 code'])


KeyError: "['WB-2 code'] not found in axis"

#### StatsCountry2 (241, 32)
ce tableau donne une description plus claire de chaque colonne. descriptions récupérées autre part

| Colonne | Description | % de NA |
|----------|------------|---------|
| Country Code | Code ISO à 3 lettres du pays (clé de jointure principale). | 0.00 % |
| Short Name | Nom court du pays. | 0.00 % |
| Table Name | Nom utilisé dans les tableaux officiels de la base de données. | 0.00 % |
| Long Name | Nom officiel complet du pays. | 0.00 % |
| 2-alpha code | Code ISO à 2 lettres du pays. | 1.24 % |
| Currency Unit | Devise officielle du pays. | 10.79 % |
| Region | Région géographique (ex : Europe & Central Asia). | 11.20 % |
| Income Group | Groupe de revenu selon la classification Banque mondiale. | 11.20 % |
| National accounts base year | Année de base utilisée pour les comptes nationaux. | 14.94 % |
| National accounts reference year | Année de référence des comptes nationaux. | 86.72 % |
| SNA price valuation | Méthode d’évaluation des prix selon le Système de Comptabilité Nationale (SNA). | 18.26 % |
| Lending category | Catégorie de prêt attribuée par la Banque mondiale. | 40.25 % |
| Other groups | Autres regroupements économiques ou administratifs. | 75.93 % |
| System of National Accounts | Version du système de comptabilité nationale utilisée. | 10.79 % |
| Alternative conversion factor | Facteur alternatif de conversion monétaire. | 80.50 % |
| PPP survey year | Année de l’enquête utilisée pour le calcul des PPA. | 39.83 % |
| Balance of Payments Manual in use | Version du manuel de balance des paiements utilisée. | 24.90 % |
| External debt Reporting status | Statut de déclaration de la dette extérieure. | 48.55 % |
| System of trade | Système de comptabilisation du commerce (général/spécial). | 17.01 % |
| Government Accounting concept | Concept comptable utilisé par le gouvernement. | 33.20 % |
| IMF data dissemination standard | Standard de diffusion des données du FMI respecté. | 24.90 % |
| Latest population census | Année du dernier recensement de la population. | 11.62 % |
| Latest household survey | Année de la dernière enquête auprès des ménages. | 41.49 % |
| Source of most recent Income and expenditure data | Source des dernières données revenus/dépenses. | 33.61 % |
| Vital registration complete | Indique si l’enregistrement des naissances/décès est complet. | 53.94 % |
| Latest agricultural census | Année du dernier recensement agricole. | 41.08 % |
| Latest industrial data | Année des dernières données industrielles disponibles. | 55.60 % |
| Latest trade data | Année des dernières données commerciales disponibles. | 23.24 % |
| Latest water withdrawal data | Année des dernières données sur les prélèvements d’eau. | 25.73 % |
| Unnamed: 31 | Colonne vide issue de l’export (à supprimer). | 100.00 % |


#### Dataset 3

In [ ]:
pd.set_option('display.max_rows', None)
StatsData3.isna().sum()/ len(StatsData3) * 100

#Le nombre de NA dans les colonnes de StatsData3 est très élevé, en particulier dans les colonnes "FootnoteCodes" et "ValueFootnoteSymbol", ce qui peut poser des problèmes pour l'analyse des données. Il serait important de comprendre pourquoi


Country Name        0.000000
Country Code        0.000000
Indicator Name      0.000000
Indicator Code      0.000000
1970               91.849639
1971               95.993258
1972               95.984012
1973               95.992356
1974               95.971497
1975               90.156382
1976               95.773849
1977               95.763589
1978               95.763364
1979               95.849842
1980               89.951631
1981               95.627953
1982               95.770692
1983               95.663694
1984               95.647233
1985               89.819264
1986               95.560867
1987               95.643286
1988               95.653321
1989               95.767422
1990               85.973527
1991               91.607342
1992               91.482642
1993               91.454455
1994               91.266278
1995               85.189248
1996               91.340128
1997               91.718287
1998               90.426076
1999               86.601085
2000          

In [179]:
StatsData3.dtypes


Country Name          str
Country Code          str
Indicator Name        str
Indicator Code        str
1970              float64
1971              float64
1972              float64
1973              float64
1974              float64
1975              float64
1976              float64
1977              float64
1978              float64
1979              float64
1980              float64
1981              float64
1982              float64
1983              float64
1984              float64
1985              float64
1986              float64
1987              float64
1988              float64
1989              float64
1990              float64
1991              float64
1992              float64
1993              float64
1994              float64
1995              float64
1996              float64
1997              float64
1998              float64
1999              float64
2000              float64
2001              float64
2002              float64
2003              float64
2004        

In [ ]:
StatsData3.dtypes

In [ ]:
z == 4

In [ ]:
z